### ISSUE TITLE
readkey() does not work with esc key

### ISSUE THREAD

**Issue by ids1024** [magmax/python-readchar#2]
You may already know this, but when escape is pressed, readkey() does not return.  I understand why this is (some other keys are escape sequences), but there must be a way to handle this, since many programs seem to be able to do it.

Edit: Additionally, it looks like Windows doesn't use ANSI escape codes, so readkey might not work with non-ascii keys on Windows. I have not tested that.

**Comment by magmax** [comment]
Hmmm... Interesting.

I really have no idea how to implement that. If you see the readchar_linux.py code, everything reduces to read from stdin (just `sys.stdin.read(1)`). The problem with ESC key is the code it generates: 0x21, just the first code generated by F1-F12 or any ALT/CTRL combination.

How can it be implemented? Well... I find three ways:
- by adding a parameter to return the ESC code when read. If it is set to true, the ESC management is delegated to the caller. I really hate this solution.
- by checking if there is any character in stdin, ready to be read. This is the best one, but there is no primitive to do that.
- by creating two threads. The first one will read the char and the other will give some time to get the char. If no additional character is ready by then, then return just the ESC key.

The last solution may fit the bill, but it can generate unexpected results: it could return different results depending on the time it requires to be generated. And it may add an important delay (just a few milliseconds but for each `readkey` call).

I really do not want to do that, so I think I will leave this issue as open, so anyone can suggest another solution or implementation.

Thank you very much.

**Comment by jsoriano** [comment]
Generally speaking, in UNIX systems you could use `select` to check if there is something else to read.
The problem in python is that file objects are usually buffered, e.g. `sys.stdin` is. So when reading from them, even if you only want to read a character, it will read its buffer size, in the next reads it will be giving you characters from its buffer, not from the underlying file itself.
This provokes `select` (that in python is directly mapped to the system call and doesn't know anything about python buffers) to see the OS buffer empty, because python has already buffered everything, so for it there is nothing to read.

This can be workarounded in two ways:
* Running python unbuffered (-u flag).
* Programatically by "re-opening" stdin without buffer, see an example bellow:

```
import sys
import select
import termios
import tty
import os

old_settings = termios.tcgetattr(sys.stdin)
tty.setcbreak(sys.stdin.fileno())
try:
    unbuffered = os.fdopen(sys.stdin.fileno(), 'r', 0)
    while True:
        if select.select([unbuffered,],[],[],0.0)[0]:
            print ord(unbuffered.read(1))
            sys.stdout.flush()
finally:
    termios.tcsetattr(sys.stdin, termios.TCSADRAIN, old_settings)
```

I hope it helps.

**Comment by ids1024** [comment]
Another way to get around the buffering is to use the lower level os.read.

**Comment by jsoriano** [comment]
True, this is probably a better way to do it.

**Comment by magmax** [comment]
I will try this evening.

**Comment by magmax** [comment]
published in version 1.0.0.

Thank you!

**Comment by sieira** [comment]
This seems not to be working anymore

```py
In [1]: from readchar import readkey

In [2]: readkey()
Out[2]: '\x1b\x1b'
```

I had to press twice


```py
In [3]: import readchar

In [4]: readchar.__version__
Out[4]: '0.0.8'
```

```sh
% pip freeze | grep readchar
-e git://github.com/magmax/python-readchar@84ca04ab4992f61dfa6a911ea6358c10afd7af57#egg=readchar
```
